GHC 8.0 support

GHC 8 is just around the corner, so we should support it the sooner the better.

I added `ghc < 8` upper bounds to affected `hlint` releases (.e.g https://hackage.haskell.org/package/hint-0.5.1/revisions/). Example failure:

```
/opt/ghc/8.0.1/bin/ghc --make -fbuilding-cabal-package -O -static -dynamic-too -dynosuf dyn_o -dynhisuf dyn_hi -outputdir dist/dist-sandbox-ee3ee364/build -odir dist/dist-sandbox-ee3ee364/build -hidir dist/dist-sandbox-ee3ee364/build -stubdir dist/dist-sandbox-ee3ee364/build -i -idist/dist-sandbox-ee3ee364/build -isrc -idist/dist-sandbox-ee3ee364/build/autogen -Idist/dist-sandbox-ee3ee364/build/autogen -Idist/dist-sandbox-ee3ee364/build -optP-include -optPdist/dist-sandbox-ee3ee364/build/autogen/cabal_macros.h -this-unit-id hint-0.5.1-IH6raNB1YVmFA2FWK6N0tQ -hide-all-packages -no-user-package-db -package-db /tmp/extra16854636236/.cabal-sandbox/x86_64-linux-ghc-8.0.0.20160329-packages.conf.d -package-db dist/dist-sandbox-ee3ee364/package.conf.inplace -package-id base-4.9.0.0 -package-id directory-1.2.5.1 -package-id exceptions-0.8.2.1-KRhvi1lhaFUEiXnCWPbwYD -package-id filepath-1.4.1.0 -package-id ghc-8.0.0.20160329 -package-id ghc-paths-0.1.0.9-GIOnKzk0HmEBZ77Q1HsThK -package-id mtl-2.2.1-6qsR1PHUy5lL47Hpoa4jCM -package-id random-1.1-Du7zLOcytk94qoeAeKnxmZ -package-id unix-2.7.2.0 -XHaskell98 -XCPP -XGeneralizedNewtypeDeriving -XDeriveDataTypeable -XMagicHash -XFunctionalDependencies -XRank2Types -XScopedTypeVariables -XExistentialQuantification Language.Haskell.Interpreter Language.Haskell.Interpreter.Extension Language.Haskell.Interpreter.Unsafe Hint.GHC Hint.Base Hint.InterpreterT Hint.Compat Hint.CompatPlatform Hint.Configuration Hint.Extension Hint.Context Hint.Conversions Hint.Eval Hint.Parsers Hint.Reflection Hint.Typecheck Hint.Util Hint.Annotations Control.Monad.Ghc -Wall
[ 1 of 19] Compiling Hint.Util        ( src/Hint/Util.hs, dist/dist-sandbox-ee3ee364/build/Hint/Util.o )
[ 2 of 19] Compiling Hint.CompatPlatform ( src/Hint/CompatPlatform.hs, dist/dist-sandbox-ee3ee364/build/Hint/CompatPlatform.o )
[ 3 of 19] Compiling Control.Monad.Ghc ( src/Control/Monad/Ghc.hs, dist/dist-sandbox-ee3ee364/build/Control/Monad/Ghc.o )

src/Control/Monad/Ghc.hs:28:1: warning: [-Wredundant-constraints]
    • Redundant constraints: (Functor m, MonadCatch m)
    • In the type signature for:
           runGhcT :: (Functor m, MonadIO m, MonadCatch m, MonadMask m) =>
                      Maybe FilePath -> GhcT m a -> m a
[ 4 of 19] Compiling Hint.GHC         ( src/Hint/GHC.hs, dist/dist-sandbox-ee3ee364/build/Hint/GHC.o )
[ 5 of 19] Compiling Hint.Compat      ( src/Hint/Compat.hs, dist/dist-sandbox-ee3ee364/build/Hint/Compat.o )
[ 6 of 19] Compiling Hint.Extension   ( src/Hint/Extension.hs, dist/dist-sandbox-ee3ee364/build/Hint/Extension.o )
[ 7 of 19] Compiling Language.Haskell.Interpreter.Extension ( src/Language/Haskell/Interpreter/Extension.hs, dist/dist-sandbox-ee3ee364/build/Language/Haskell/Interpreter/Extension.o )
[ 8 of 19] Compiling Hint.Base        ( src/Hint/Base.hs, dist/dist-sandbox-ee3ee364/build/Hint/Base.o )

src/Hint/Base.hs:44:5: warning: [-Wredundant-constraints]
    • Redundant constraint: Functor n
    • In the type signature for:
           runGhc :: (MonadIO n, MonadMask n, Functor n) => GHC.GhcT n a
      In the ambiguity check for ‘runGhc’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      When checking the class method:
        runGhc :: forall (m :: * -> *).
                  MonadInterpreter m =>
                  forall a. RunGhc m a
[ 9 of 19] Compiling Hint.Annotations ( src/Hint/Annotations.hs, dist/dist-sandbox-ee3ee364/build/Hint/Annotations.o )
[10 of 19] Compiling Hint.Configuration ( src/Hint/Configuration.hs, dist/dist-sandbox-ee3ee364/build/Hint/Configuration.o )
[11 of 19] Compiling Hint.Reflection  ( src/Hint/Reflection.hs, dist/dist-sandbox-ee3ee364/build/Hint/Reflection.o )
[12 of 19] Compiling Hint.Conversions ( src/Hint/Conversions.hs, dist/dist-sandbox-ee3ee364/build/Hint/Conversions.o )
[13 of 19] Compiling Hint.Context     ( src/Hint/Context.hs, dist/dist-sandbox-ee3ee364/build/Hint/Context.o )
[14 of 19] Compiling Hint.InterpreterT ( src/Hint/InterpreterT.hs, dist/dist-sandbox-ee3ee364/build/Hint/InterpreterT.o )

src/Hint/InterpreterT.hs:178:46: error:
    • Couldn't match expected type ‘GHC.SrcSpan’
                  with actual type ‘GHC.Severity’
    • In the second argument of ‘mkGhcError’, namely ‘src’
      In the expression: mkGhcError renderErrMsg src style msg
      In an equation for ‘errorEntry’:
          errorEntry = mkGhcError renderErrMsg src style msg

src/Hint/InterpreterT.hs:178:50: error:
    • Couldn't match expected type ‘GHC.PprStyle’
                  with actual type ‘GHC.SrcSpan’
    • In the third argument of ‘mkGhcError’, namely ‘style’
      In the expression: mkGhcError renderErrMsg src style msg
      In an equation for ‘errorEntry’:
          errorEntry = mkGhcError renderErrMsg src style msg

src/Hint/InterpreterT.hs:178:56: error:
    • Couldn't match type ‘GHC.PprStyle’ with ‘GHC.SDoc’
      Expected type: GHC.Message
        Actual type: GHC.PprStyle
    • In the fourth argument of ‘mkGhcError’, namely ‘msg’
      In the expression: mkGhcError renderErrMsg src style msg
      In an equation for ‘errorEntry’:
          errorEntry = mkGhcError renderErrMsg src style msg

src/Hint/InterpreterT.hs:179:8: error:
    • Couldn't match expected type ‘GHC.MsgDoc -> IO ()’
                  with actual type ‘IO ()’
    • Possible cause: ‘modifyIORef’ is applied to too many arguments
      In the expression: modifyIORef r (errorEntry :)
      In the expression:
        let
          renderErrMsg = GHC.showSDoc df
          errorEntry = mkGhcError renderErrMsg src style msg
        in modifyIORef r (errorEntry :)
      In an equation for ‘mkLogHandler’:
          mkLogHandler r df _ src style msg
            = let
                renderErrMsg = GHC.showSDoc df
                errorEntry = mkGhcError renderErrMsg src style msg
              in modifyIORef r (errorEntry :)
xcabal: Error: some packages failed to install:
hint-0.5.1 failed during the building phase. The exception was:
ExitFailure 1
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -6,6 +6,8 @@ matrix:
       addons: {apt: {packages: [cabal-install-1.18, ghc-7.8.4],  sources: [hvr-ghc]}}
     - env: CABALVER=1.22 GHCVER=7.10.3
       addons: {apt: {packages: [cabal-install-1.22, ghc-7.10.3], sources: [hvr-ghc]}}
+    - env: CABALVER=1.24 GHCVER=8.0.1
+      addons: {apt: {packages: [cabal-install-1.24, ghc-8.0.1], sources: [hvr-ghc]}}
 
 before_install:
   - export PATH=/opt/ghc/$GHCVER/bin:/opt/cabal/$CABALVER/bin:$PATH

--- a/hint.cabal
+++ b/hint.cabal
@@ -45,7 +45,7 @@ Test-Suite unit-tests
 
 Library
   build-depends: base == 4.*,
-                 ghc >= 7.6 && < 8,
+                 ghc >= 7.6 && < 8.2,
                  ghc-paths,
                  mtl,
                  filepath,

--- a/src/Hint/InterpreterT.hs
+++ b/src/Hint/InterpreterT.hs
@@ -182,7 +182,11 @@ newSessionData a =
        }
 
 mkLogHandler :: IORef [GhcError] -> GhcErrLogger
+#if __GLASGOW_HASKELL__ >= 800
+mkLogHandler r df _ _ src style msg =
+#else
 mkLogHandler r df _ src style msg =
+#endif
     let renderErrMsg = GHC.showSDoc df
         errorEntry = mkGhcError renderErrMsg src style msg
     in modifyIORef r (errorEntry :)

--- a/src/Hint/Parsers.hs
+++ b/src/Hint/Parsers.hs
@@ -7,6 +7,7 @@ import Hint.Base
 import Control.Monad.Trans (liftIO)
 
 import qualified Hint.GHC as GHC
+import qualified DynFlags as GHC
 
 data ParseResult = ParseOk | ParseError GHC.SrcSpan GHC.Message
 
@@ -50,7 +51,11 @@ failOnParseError parser expr = mayFail go
                              dflags <- runGhc GHC.getSessionDynFlags
                              let logger'  = logger dflags
                                  errStyle = GHC.defaultErrStyle dflags
-                             liftIO $ logger' GHC.SevError
+                             liftIO $ logger'
+#if __GLASGOW_HASKELL__ >= 800
+                                              GHC.NoReason
+#endif
+                                              GHC.SevError
                                               span
                                               errStyle
                                               err

